File: static/js/components/Admin/AdminDashBoard.js
export default Vue.component("admin-dashboard", {
    template: `
      <v-container>
         <v-btn color="danger" @click="downloadReport">Download Report</v-btn>
        <edit-service-form
        v-if="editingService || addingService"
        :service="editingService || newService"
        @service-saved="refreshServices"
        @cancel="cancelEdit"
      ></edit-service-form>

        <!-- Services Table -->
        <div v-else>
        <v-card class="mt-4">
          <v-card-title>
            <h2>Services</h2>
            <v-spacer></v-spacer>
            <v-btn color="primary" @click="addService">Add New Service</v-btn>
          </v-card-title>

          <v-data-table
            :headers="serviceHeaders"
            :items="services"
            item-value="id"
            class="elevation-1"
            dense
          >
           <template v-slot:item.action="{ item }" @click="showService(item.id)">

              <v-btn small color="blue" @click="editService(item.id)"><v-icon>mdi-pencil</v-icon></v-btn>
             
              <v-btn small color="red" @click="deleteService(item.id)"><v-icon>mdi-delete</v-icon></v-btn>
            </template>
          </v-data-table>
        </v-card>

     
  
        <v-card class="mt-4">
          <v-card-title>
            <h2>Professionals</h2>
          </v-card-title>
          <v-data-table
            :headers="professionalHeaders"
            :items="professionals"
            item-value="id"
            class="elevation-1"
            dense
          >
           <template v-slot:item.action="{ item }">
           
           <v-btn 
           small 
           color="green" 
           :disabled="item.is_active" 
           @click="toggleStatus(item.id, true)"
         >
           Approve
         </v-btn>
         <v-btn 
           small 
           color="orange" 
           :disabled="!item.is_active" 
           @click="toggleStatus(item.id, false)"
         >
           Reject
         </v-btn>
              <v-btn small color="red" @click="deleteProfessional(item.id)">Delete</v-btn>
              <v-btn small color="blue" @click="viewDocument(item.attached_docs_path)" > View  </v-btn>
            </template>
          </v-data-table>
        </v-card>
  
        <!-- Service Requests Table -->
        <v-card class="mt-4">
          <v-card-title>
            <h2>Service Requests</h2>
          </v-card-title>
          <v-data-table
            :headers="serviceRequestHeaders"
            :items="serviceRequests"
            item-value="id"
            class="elevation-1"
            dense
          >
          <template v-slot:item.action="{ item }">
          <v-btn 
           small 
           color="red" 
           :disabled="!item.active"
           @click="toggleStatus(item.id, false)"
         >
           Block
         </v-btn>
          </template>
          </v-data-table>
        </v-card>
        </div>
      </v-container>
    `,
    data() {
      return {
        services: [],
        professionals: [],
        serviceRequests: [],
        serviceHeaders: [
          { text: "ID", value: "id" },
          { text: "Name", value: "name" },
          { text: "Base Price", value: "base_price" },
          { text: "Action", value: "action", sortable: false },
        ],
        professionalHeaders: [
          { text: "ID", value: "id" },
          { text: "Name", value: "name" },
          { text: "Service Type", value: "service_type" },
          { text: "Experience (Years)", value: "experience_years" },
          { text: "Rating", value: "rating" },
          { text: "Active", value: "is_active" },
          { text: "Action", value: "action", sortable: false },
      
        ],
        serviceRequestHeaders: [
          { text: "ID", value: "id" },
          { text: "Status", value: "status" },
          { text: "Customer ID", value: "customer_id" },
          { text: "Professional ID", value: "professional_id" },
          { text: "Date of Request", value: "date_of_request" },
          { text: "Remarks", value: "remarks" },
          { text: "Action", value: "action", sortable: false },
        ],
        editingService: null,
        addingService: false,
        newService: {
        name: "",
        description: "",
        base_price: 0,
        time_required: 0,
      },
      };

    },
    methods: {
      async fetchDashboardData() {
        try {
          const response = await fetch("/api/admin-dashboard", {
            method: "GET",
            headers: { "authentication-token": localStorage.getItem("token") },
          });
          const data = await response.json();
          this.services = data.services;
          this.professionals = data.professionals;
          this.serviceRequests = data.service_requests;
        } catch (error) {
          console.error("Failed to fetch dashboard data:", error);
          alert("An error occurred while fetching data.");
        }
      },
      addService() {
        this.addingService = true; 
      },
      async editService(serviceId) {
        try {
          const response = await fetch(`/api/service/${serviceId}`);
          const fullServiceDetails = await response.json();
          
          if (response.ok) {
            this.editingService = fullServiceDetails;
           
            
          } else {
            alert(fullServiceDetails.message || "Failed to fetch service details.");
          }
        } catch (error) {
          console.error("Error fetching service details:", error);
          alert("An error occurred.");
        }
      },
      cancelEdit() {
        this.editingService = null; 
        this.addingService = false;
        this.fetchDashboardData()
      },
      refreshServices() {
        this.cancelEdit();
        this.fetchDashboardData();
      },
      deleteService(serviceId) {
        if (confirm("Are you sure you want to delete this service?")) {
          fetch(`/api/service/${serviceId}`, {
            method: "DELETE",
          })
            .then((response) => response.json())
            .then((result) => {
              if (result.success) {
                this.services = this.services.filter(
                  (service) => service.id !== serviceId
                );
                alert("Service deleted successfully.");
              } else {
                alert(result.message || "Failed to delete service.");
              }
            })
            .catch((error) => {
              console.error("Error deleting service:", error);
              alert("An error occurred.");
            });
        }
      },
      async viewDocument(filePath) {
        fetch(`${filePath}`) 
        .then(response => {
          if (!response.ok) {
            throw new Error('Network response was not ok');
          }
          return response.blob(); 
        })
        .then(blob => {
          const fileUrl = URL.createObjectURL(blob); 
          window.open(fileUrl, '_blank'); 
        })
        .catch(error => {
          console.error('Error fetching the file:', error);
        });
      },
      async toggleStatus(userId, approve) {
        try {
          const response = await fetch("/api/change-status", {
            method: "PUT",
            headers: {
              "Content-Type": "application/json",
              "authentication-token": localStorage.getItem("token"),
            },
            body: JSON.stringify({ user_id: userId }),
          });
    
          const result = await response.json();
          if (response.ok) {
          
            const user = this.professionals.find((p) => p.id === userId);
            if (user) {
              user.is_active = approve; 
            }
            alert("Status changed")
      
          } else {
            alert(result.message || "Failed to toggle status.");
          }
        } catch (error) {
          console.error("Error toggling status:", error);
          alert("An error occurred.");
        }
      },
      async deleteProfessional(profid){
        if (confirm("Are you sure you want to delete?")) {
          fetch(`/api/delete-professional/${profid}`, {
            method: "DELETE",
          })
            .then((response) => response.json())
            .then((result) => {
              if (result.success) {
                this.professionals = this.professionals.filter(
                  (prof) => prof.id !== profid
                );
                alert(" deleted successfully.");
                this.fetchDashboardData();
              } else {
                alert(result.message || "Failed to delete");
              }
            })
            .catch((error) => {
              console.error("Error deleting :", error);
              alert("An error occurred.");
            });
        }

      },
      async downloadReport(){
        fetch('/download-csv')
        .then((response)=>response.json())
        .then((result)=>{
           if(result.success){
            task_id=result.task_id
           }
        }).catch((error) => {
          console.error("Error deleting :", error);
          alert("An error occurred.");
        });

        fetch(`get-csv/${task_id}`)
        .then((response)=>response.json())
        .then((result)=>{
          if(result.success){
            const blob=result.blob()
            const fileUrl = window.URL.createObjectURL(blob);
            const link = document.createElement("a");
            link.href = fileUrl;
            link.download = "result.csv"; 
            document.body.appendChild(link);
            link.click(); 
            document.body.removeChild(link); 
            window.URL.revokeObjectURL(fileUrl); 
          }
        }).catch((error) => {
          console.error("Error deleting :", error);
          alert("An error occurred.");
        });

      }
    },
    components: {
      EditServiceForm: () => import("./EditService.js"),
    },
    created() {
      this.fetchDashboardData();
    },
  });

File: applications/resources/adminDashboard.py
from flask_restful import Resource
from flask_security import auth_required, roles_required
from applications.database.models import Service, ProfessionalDetails, ServiceRequest,User,db
from flask import jsonify 
from flask_restful import reqparse
from applications.instance import cache




class AdminDashboardResource(Resource):
    @auth_required('token')
    @roles_required("Admin")
    @cache.cached(timeout=300) 
    def get(self):
      
        services = Service.query.all()
        professionals = ProfessionalDetails.query.all()
        service_requests = ServiceRequest.query.all()

        # Serialize data for JSON response
        service_data = [
            {"id": service.id, "name": service.name, "base_price": service.base_price}
            for service in services
        ]
        
        professional_data = [
            {
                "id": pro.user_id,
                "name": pro.user.name,  
                "service_type": pro.service_type.name,  
                "experience_years": pro.experience_years,
                "rating": pro.rating,
                "is_active": pro.is_active,
                "attached_docs_path":pro.attached_docs_path,

            }
            for pro in professionals
        ]
        
        service_request_data = [
            {
                "id": request.id,
                "status": request.status,
                "customer_id": request.customer_id,
                "professional_id": request.professional_id,
                "date_of_request": request.date_of_request,
                "remarks": request.remarks,
                "active": request.customer.active,
            }
            for request in service_requests
        ]

        return jsonify({
            "services": service_data,
            "professionals": professional_data,
            "service_requests": service_request_data
        })
    

    @auth_required("token")
    @roles_required("Admin")
    def put(self):
        
        parser = reqparse.RequestParser()
        parser.add_argument("user_id", type=int, required=True, help="User ID is required.")
        args = parser.parse_args()
        
        user_id = args["user_id"]

   
        user = User.query.get(user_id)
        
        if not user:
            return {"message": "User not found"}, 404

      
        professional_details = ProfessionalDetails.query.filter_by(user_id=user_id).first()
        

        if professional_details:
       
            professional_details.is_active = not professional_details.is_active
            db.session.commit()
            cache.clear()
            
            return {"message": f"Professional status toggled to {professional_details.is_active}"}, 200
        else:
            user.active = not user.active
            db.session.commit()
            cache.clear()

            return {"message": f"Customer status toggled to {user.active}"}, 200
        
   
    @auth_required("token")
    @roles_required("Admin")
    def delete(self, professional_id):
       
        professional = ProfessionalDetails.query.filter_by(user_id=professional_id).first()
        
        if not professional:
            return {"message": "Professional not found."}, 404
        user = User.query.filter_by(id=professional.user_id).first()
        
        if not user:
            return {"message": "Associated user not found."}, 404

        try:
    

            db.session.delete(professional)
            db.session.delete(user)
            db.session.commit()
            cache.clear()

            return {"message": "Professional and associated user details deleted successfully."}, 200
        except Exception as e:
            db.session.rollback()
            return {"message": f"An error occurred: {str(e)}"}, 500